### PR TITLE
UAR-900: Tell us about trust JSON data

### PR DESCRIPTION
**Pull request by joshuahigginsonCH**
### JIRA link
https://companieshouse.atlassian.net/browse/UAR-900

### Change description
https://companieshouse.atlassian.net/wiki/spaces/ROE2/pages/4234608689/Notes+for+ROE+Api+App+to+Add+Trusts+to+Update+JSON+OE02

Testing screenshots and steps:

- Ran through Update flow on UI
- Hit the filings endpoint locally

Notes:
- Assumptions for further tickets look correct

Trust data returned:
Screenshot 2023-07-21 at 13 41 11

Additions flow:
- Trusts added
Screenshot 2023-07-21 at 13 39 15

- No trusts added
Screenshot 2023-07-21 at 13 39 43

- Additions are still being checked

### Work checklist

- [] Tests added where applicable

### Merge instructions

We are committed to keeping commit history clean, consistent and linear. To achieve this, this commit should be structured as follows:

```
<type>[optional scope]: <description>
```

and contain the following structural elements:

- fix: a commit that patches a bug in your codebase (this correlates with PATCH in semantic versioning),
- feat: a commit that introduces a new feature to the codebase (this correlates with MINOR in semantic versioning),
- BREAKING CHANGE: a commit that has a footer `BREAKING CHANGE:` introduces a breaking API change (correlating with MAJOR in semantic versioning). A BREAKING CHANGE can be part of commits of any type,
- types other than `fix:` and `feat:` are allowed, for example `build:`, `chore:`, `ci:`, `docs:`, `style:`, `refactor:`, `perf:`, `test:`, and others,
- footers other than `BREAKING CHANGE: <description>` may be provided.

### CODE REVIEW

File: src/main/java/uk/gov/companieshouse/overseasentitiesapi/service/BeneficialOwnerAdditionService.java [hunk 1]
```diff
@@ -89,6 +89,10 @@ private IndividualBeneficialOwnerAddition getIndividualBeneficialOwnerAddition(
                         addressDtoToAddress(serviceAddress),
                         natureOfControls,
                         isOnSanctionsList);
+        
+        if (!bo.getTrustIds().isEmpty()) {
+                individualBeneficialOwnerAddition.setTrustIds(bo.getTrustIds());
```
Comment (AndreKramer7654) [inline src/main/java/uk/gov/companieshouse/overseasentitiesapi/service/BeneficialOwnerAdditionService.java:94]: Indentation?

File: src/main/java/uk/gov/companieshouse/overseasentitiesapi/service/BeneficialOwnerAdditionService.java [hunk 2]
```diff
@@ -127,6 +131,10 @@ private CorporateEntityBeneficialOwnerAddition getCorporateEntityBeneficialOwner
 
         corporateEntityBeneficialOwnerAddition.setCorporateName(bo.getName());
 
+        if (!bo.getTrustIds().isEmpty()) {
+                corporateEntityBeneficialOwnerAddition.setTrustIds(bo.getTrustIds());
```
Comment (AndreKramer7654) [inline src/main/java/uk/gov/companieshouse/overseasentitiesapi/service/BeneficialOwnerAdditionService.java:135]: Indentation? Maybe always just set - which woul dbe set to null if getTrustIds is empty.

File: src/main/java/uk/gov/companieshouse/overseasentitiesapi/service/BeneficialOwnerChangeService.java [hunk 3]
```diff
@@ -225,6 +225,10 @@ private BeneficialOwnerChange<CorporateBeneficialOwnerPsc> convertBeneficialOwne
         Identification::getRegistrationNumber,
         CompanyIdentification::setRegistrationNumber);
 
+    if (!beneficialOwnerCorporateDto.getTrustIds().isEmpty()) {
```
Comment (AndreKramer7654) [inline src/main/java/uk/gov/companieshouse/overseasentitiesapi/service/BeneficialOwnerChangeService.java:228]: Maybe always just set - which would be set to null if getTrustIds is empty.

**Comment by AndreKramer7654** [comment]
For all changed BOs/PSCs. We need to force a change to be outputted if the BO/PSC has any trusts. Somewhere in the BenficialOwnerChangeService. Use case: the reviewed BO has not changed (any info fields) but the user has added in trust info. Possibly she just re-entered the trusts that were there before in registration but we need to create a "change" to list the added trusts.

File: src/main/java/uk/gov/companieshouse/overseasentitiesapi/service/BeneficialOwnerChangeService.java (hunk 3 above)
Comment (joshuahigginsonCH) [reply src/main/java/uk/gov/companieshouse/overseasentitiesapi/service/BeneficialOwnerChangeService.java:228]: yep i have added this just was waiting to push with unit tests

File: src/main/java/uk/gov/companieshouse/overseasentitiesapi/service/BeneficialOwnerAdditionService.java (hunk 1 above)
Comment (joshuahigginsonCH) [reply src/main/java/uk/gov/companieshouse/overseasentitiesapi/service/BeneficialOwnerAdditionService.java:94]: Indentation fixed, unsure what you mean by the question

Comment (AndreKramer7654) [reply src/main/java/uk/gov/companieshouse/overseasentitiesapi/service/BeneficialOwnerAdditionService.java:94]: Consistent now but I was questioning the indentation of the whole file. And commenting that we could just always set here but ok.

File: src/main/java/uk/gov/companieshouse/overseasentitiesapi/service/BeneficialOwnerAdditionService.java [hunk 4]
```diff
@@ -89,6 +89,10 @@ private IndividualBeneficialOwnerAddition getIndividualBeneficialOwnerAddition(
                         addressDtoToAddress(serviceAddress),
                         natureOfControls,
                         isOnSanctionsList);
+        
+        if (!bo.getTrustIds().isEmpty()) {
```
Comment (AndreKramer7654) [inline src/main/java/uk/gov/companieshouse/overseasentitiesapi/service/BeneficialOwnerAdditionService.java:93]: We should protect against bo.getTrustIds() returning null. 
if (bo.getTrustIds() != null && !bo.getTrustIds().isEmpty()) {

File: src/main/java/uk/gov/companieshouse/overseasentitiesapi/service/BeneficialOwnerChangeService.java [hunk 5]
```diff
@@ -225,6 +225,11 @@ private BeneficialOwnerChange<CorporateBeneficialOwnerPsc> convertBeneficialOwne
...
```
Comment (AndreKramer7654) [inline src/main/java/uk/gov/companieshouse/overseasentitiesapi/service/BeneficialOwnerChangeService.java:228]: We should protect against bo.getTrustIds() returning null.

File: src/main/java/uk/gov/companieshouse/overseasentitiesapi/service/BeneficialOwnerChangeService.java [hunk 6]
```diff
@@ -397,6 +402,11 @@ private BeneficialOwnerChange<IndividualBeneficialOwnerPsc> convertBeneficialOwn
         IndividualBeneficialOwnerPsc::setPersonName
     );
 
+    if (!beneficialOwnerIndividualDto.getTrustIds().isEmpty()) {
```
Comment (AndreKramer7654) [inline src/main/java/uk/gov/companieshouse/overseasentitiesapi/service/BeneficialOwnerChangeService.java:405]: We should protect against bo.getTrustIds() returning null.

File: src/test/java/uk/gov/companieshouse/overseasentitiesapi/service/BeneficialOwnerAdditionServiceTest.java [hunk 7]
```diff
@@ -223,7 +225,7 @@ private List<BeneficialOwnerCorporateDto> getCorporateBeneficialOwners() {
         corporateBeneficialOwner.setPublicRegisterName("Register name");
         corporateBeneficialOwner.setRegistrationNumber("Registration number");
         corporateBeneficialOwner.setOnSanctionsList(true);
-
+        corporateBeneficialOwner.setTrustIds(new ArrayList<>());
```
Comment (AndreKramer7654) [inline src/test/java/uk/gov/companieshouse/overseasentitiesapi/service/BeneficialOwnerAdditionServiceTest.java:228]: test for null as well

File: src/main/java/uk/gov/companieshouse/overseasentitiesapi/model/updatesubmission/UpdateSubmission.java [hunk 8]
```diff
@@ -62,6 +64,9 @@ public class UpdateSubmission {
     @JsonProperty(CESSATIONS_FIELD)
     private List<Cessation> cessations;
 
+    @JsonProperty(TRUSTS_ADDITIONS_FIELD)
```
Comment (AndreKramer7654) [inline src/main/java/uk/gov/companieshouse/overseasentitiesapi/model/updatesubmission/UpdateSubmission.java:67]: @JsonInclude(NON_NULL)

File: src/main/java/uk/gov/companieshouse/overseasentitiesapi/service/FilingsService.java [hunk 9]
```diff
@@ -249,6 +249,7 @@ private void setUpdateSubmissionData(
       data.put(CHANGES_FIELD, updateSubmission.getChanges());
       data.put(ADDITIONS_FIELD, updateSubmission.getAdditions());
       data.put(CESSATIONS_FIELD, updateSubmission.getCessations());
```
Comment (AndreKramer7654) [inline src/main/java/uk/gov/companieshouse/overseasentitiesapi/service/FilingsService.java:251]: if (updateSubmission.getTrustAdditions() != null && ! updateSubmission.getTrustAdditions().isEmpty) {
     data.put(TRUST_DATA, updateSubmission.getTrustAdditions());
}

**Comment by AndreKramer7654** [comment]
I'd like to be sure that we handle the lists being null to lessen risk in adding this without an explicit feature flag.

**Comment by AndreKramer7654** [comment]
For the test coverage - you need some unit tests that include non-empty lists of trust ids (for Addition and Change BO services).

File: src/main/java/uk/gov/companieshouse/overseasentitiesapi/service/FilingsService.java (hunk 9 above)
Comment (AndreKramer7654) [reply src/main/java/uk/gov/companieshouse/overseasentitiesapi/service/FilingsService.java:251]: Still think we should add only if not null.

File: src/main/java/uk/gov/companieshouse/overseasentitiesapi/model/updatesubmission/UpdateSubmission.java [hunk 10]
```diff
@@ -155,4 +161,12 @@ public List<Cessation> getCessations() {
     public void setCessations(List<Cessation> cessations) {
         this.cessations = cessations;
     }
+
+    public List<TrustDataDto> getTrustAdditions() {
+        return trustAdditions;
+    }
+
+    public void setTrustAdditions(List<TrustDataDto> trustAdditions) {
```
Comment (AndreKramer7654) [inline src/main/java/uk/gov/companieshouse/overseasentitiesapi/model/updatesubmission/UpdateSubmission.java:169]: Is this called anywhere?

Comment (AndreKramer7654) [reply src/main/java/uk/gov/companieshouse/overseasentitiesapi/model/updatesubmission/UpdateSubmission.java:169]: Needs to be set in  collectUpdateSubmissionData() from userSubmission.

File: src/main/java/uk/gov/companieshouse/overseasentitiesapi/service/FilingsService.java (hunk 9 above)
Comment (AndreKramer7654) [reply src/main/java/uk/gov/companieshouse/overseasentitiesapi/service/FilingsService.java:255]: pushed a change

**Review by AndreKramer7654 — APPROVED** [review]

**Review by conuirchgov — APPROVED** [review]
